tests/Unit/Css/CssDocumentTest.php

```diff
@@ -437,15 +441,52 @@ public function provideCssWithoutNonConditionalAtRules(): array
      * @dataProvider provideValidAtCharsetRules
      * @dataProvider provideInvalidAtCharsetRules
      */
-    public function discardsValidOrInvalidAtCharsetRule(string $css): void
+    public function discardsValidOrInvalidAtCharsetRuleNotInDebugMode(string $css): void
     {
-        $subject = new CssDocument($css);
+        $subject = new CssDocument($css, false);
 
         $result = $subject->renderNonConditionalAtRules();
 
         self::assertSame('', $result);
     }
 
+    /**
+     * @test
+     *
+     * @param string $css
+     *
+     * @dataProvider provideValidAtCharsetRules
+     */
+    public function discardsValidAtCharsetRuleInDebugMode(string $css): void
+    {
+        $subject = self::createDebugSubject($css);
+
+        $result = $subject->renderNonConditionalAtRules();
+
+        self::assertSame('', $result);
+    }
+
+    /**
+     * @test
+     *
+     * @param string $css
+     *
+     * @dataProvider provideInvalidAtCharsetRules
+     */
+    public function throwsExceptionForOrDiscardsInvalidAtCharsetRuleInDebugMode(string $css): void
+    {
+        try {
+            $subject = self::createDebugSubject($css);
+
+            $result = $subject->renderNonConditionalAtRules();
+
+            self::assertSame('', $result);
```

Comment:
Will this line ever be reached?

Reply:
I replied at length but it's been lost because GitHub's UI/UX is rubbish and I accidentally clicked a link.

Reply:
https://github.com/MyIntervals/emogrifier/pull/1142/files#diff-2eda2d27b064a67021c7004fd070acb00ac6bf08695521484af6c087c6eec0baR1425

Reply:
In the above link it was possible to split test data up on logical grounds - one set of data syntactically invalid, the other merely with property names/values that don't conform to the spec.

Reply:
In this case there doesn't appear to be any logical split between which test data causes an exception and for which the invalid rules are simply discarded.

Reply:
We already test in non-debug mode that the invalid rules are discarded.

Reply:
In debug mode, to throw an exception in all possible cases of invalid CSS would, I think, be considered a feature request that is never likely to be honoured (very high-hanging fruit, and who is ever going to take the time to pick it?).

Reply:
Thus we have a situation where the test can pass if either of the following conditions are met:
- An exception is thrown;
- The invalid CSS rule is discarded.

Reply:
Usually in PHPUnit we can use `logicalOr` to test for this.

Reply:
But if one of the conditions is an exception being thrown, AFAIAA, this is not possible.

Reply:
Hence this construct.

Reply:
We (as the Emogrifier maintainers) have the secret knowledge of what actually happens ;-) - and I think it's fair that we decide on one possible behavior to be the desired one and test for that. Otherwise, this would make the tests really confusing to read (because the desired behavior then is not clear to the reader).

Reply:
What actually happens here is down to the behaviour of `sabberworm/php-css-parser`.  We only know what happens because we have looked into it.  We can influence it and I note you have had a number of PRs accepted.  But at the end of the day it is a 3rd party library.

Ultimately we may be able to dispense with this intermediary class and use the Sabberworm classes directly, but we are a little way off that yet.

Some invalid CSS throws an excpetion (in debug mode).  Some doesn't.  If it throws an excpetion in debug mode, that's a bonus.  If it doesn't, it's no big deal.  But we don't control that.  This relates back to #1135 - to what extent are we testing 3rd party libraries duplicitously?